server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2403,7 +2405,18 @@ class OpAndVersion {
                     try (Engine.GetResult get = engine.get(engineGet, mappingLookup, documentParser, randomSearcherWrapper())) {
                         FieldsVisitor visitor = new FieldsVisitor(true);
                         get.docIdAndVersion().reader.document(get.docIdAndVersion().docId, visitor);
-                        List<String> values = new ArrayList<>(Strings.commaDelimitedListToSet(visitor.source().utf8ToString()));
+                        List<String> values = new ArrayList<>();
+                        if (visitor.source().length() > 0) {
+                            values.addAll(
+                                new ArrayList<>(
+                                    (List<String>) XContentMapValues.extractValue(
```

Comment:
Interesting to know why this change (and the related ones below)? Is it safer than the previous way, or more explicit?

Reply:
This is because I added assertions in `Translog.Index#invariant()` to ensure that we correctly set the source xcontent type when addind index operations to the translog. I wanted to be sure to catch all cases and tests, but I had to change tests to inject correctly formatted source instead of dummy raw bytes. I think the test now is more realistic, at the expanse of a bit of complexification.